### ISSUE TITLE
use flake: [print-dev-env | flake] is not a recognized command (NixOS)

### ISSUE THREAD

**Issue by mplanchard** [nix-community/nix-direnv#137]
Note that I've got this working for now with the `NIX_BIN_PREFIX` fix mentioned below, so no rush. I'm pretty new to NixOS, so it's very possible I'm doing something wrong, and apologies for the noise if that's the case. Thanks in advance for any help you can provide!

Anyway, I've encountered an issue on NixOS where, with a simple flake and an `.envrc` containing nothing but `use flake`, I get the following output:

```
❯ direnv allow
direnv: loading ~/s/gh/mplanchard/rust-play/.envrc
direnv: using flake
error: 'print-dev-env' is not a recognised command
Try '/nix/store/0qqkw5qjdr6l4aah0ncns13ga0zynq04-nix-2.3.16/bin/nix --help' for more information.
error: 'flake' is not a recognised command
Try '/nix/store/0qqkw5qjdr6l4aah0ncns13ga0zynq04-nix-2.3.16/bin/nix --help' for more information.
direnv: renewed cache
direnv: export ~XDG_DATA_DIRS
```

I dug into the debug output after setting -x. You can see the full output of that here: https://pastebin.com/83sU5g2u

It's weird, b/c it's doing exactly what it's supposed to do, but the version of nix that is hard-coded in the `direnvrc` file (2.3.16) doesn't support those commands, or the `--extra-experimental-features` flag:

```
rust-play on  main [!?] 
❯ /nix/store/0qqkw5qjdr6l4aah0ncns13ga0zynq04-nix-2.3.16/bin/nix flake --extra-experimental-features "nix-command flakes"
error: 'flake' is not a recognised command
Try '/nix/store/0qqkw5qjdr6l4aah0ncns13ga0zynq04-nix-2.3.16/bin/nix --help' for more information.

rust-play on  main [!?] 
❯ /nix/store/0qqkw5qjdr6l4aah0ncns13ga0zynq04-nix-2.3.16/bin/nix --extra-experimental-features "nix-command flakes" flake
error: unrecognised flag '--extra-experimental-features'
Try '/nix/store/0qqkw5qjdr6l4aah0ncns13ga0zynq04-nix-2.3.16/bin/nix --help' for more information.
```

If I set `NIX_BIN_PREFIX` to the `nix-shell` or `nix` executable that is the default on my system, it works, e.g. in `.envrc`:

```
NIX_BIN_PREFIX=$(readlink $(command -v nix))
NIX_BIN_PREFIX="${NIX_BIN_PREFIX%/*}/"

NIX_BIN_PREFIX="$NIX_BIN_PREFIX" use flake
```

Here's my flake:

```nix
{
  description = "rust playground";

  inputs = {
   nixpkgs.url = "nixpkgs/nixos-unstable";
   # Provides some nice helpers for multiple system compatibility
   flake-utils.url = "github:numtide/flake-utils";
   # Provides rust and friends
   rust-overlay.url = "github:oxalica/rust-overlay";
  };

  outputs = { self, nixpkgs, flake-utils, rust-overlay }:
    # Calls the provided function for each "default system", which
    # is the standard set.
    flake-utils.lib.eachDefaultSystem
      (system:
        # instantiate the package set for the supported system, with our
        # rust overlay
        let pkgs = import nixpkgs {
              inherit system;
              overlays = [ rust-overlay.overlay ];
            };
        in
        # "unpack" the pkgs attrset into the parent namespace
        with pkgs;
        {
          devShell = mkShell {
            nativeBuildInputs = [ bashInteractive ];
            buildInputs = [
              nixFlakes
              # Read our toolchain file to determine which version of rust,
              # components, and targets to install.
              (rust-bin.fromRustupToolchainFile ./rust-toolchain.toml)
              rust-analyzer
            ];
          };
        });
}
```

And here's the nix setup in my `configuration.nix`:

```nix
  nix = {
    package = pkgs.nixUnstable;
    extraOptions = ''
      experimental-features = nix-command flakes
      keep-outputs = true
      keep-derivations = true
    '';
  };
```

And my home-manager setup for `nix-direnv`:

```nix
        direnv = {
          enable = true;
          # turn this on once I get bash config managed by home manager.
          enableBashIntegration = true;
          nix-direnv.enable = true;
        };
```

which is in a flake with these inputs:

```nix
  inputs = {
    nixpkgs.url = "nixpkgs/nixos-21.11";
    emacs-overlay.url = "github:nix-community/emacs-overlay";
    home-manager.url = "github:nix-community/home-manager";
    home-manager.inputs.nixpkgs.follows = "nixpkgs";
  };
```

I can link the full repo with all my configs if that would be useful, although it's a huge mess right now since I just switched over to NixOS last week.

**Comment by Mic92** [comment]
You also need to enable this on 21.11 in home-manager.

```
  programs.direnv.nix-direnv.enableFlakes = true;
```

**Comment by mplanchard** [comment]
Right, so I saw that in the README, and I had already tried that. When I run `nixos-rebuild switch` with that in my config, I get this error:

```
error:
       Failed assertions:
       - matthew profile: The option definition `programs.direnv.nix-direnv.enableFlakes' in `/nix/store/byly36rz37xicifh251vvgfsvam040k2-source/flake.nix' no longer has any effect; please remove it.
       Flake support is now always enabled.
```

**Comment by mplanchard** [comment]
@Mic92 any ideas as to what to do, given that configuration option fails in home-manager?

**Comment by exzombie** [comment]
I have the same problem. Is this related to #142?

**Comment by exzombie** [comment]
Solved. It seemed that there is some mismatch in nix versions between NixOS and home-manager. Turns out that it was true, and it was the reason for the home-manager direnv module rejecting the `enableFlakes` option.

@mplanchard I suggest you add `home.enableNixpkgsReleaseCheck = true;` to your home-manager config to see if you have the same issue. The releases of nixpkgs and home-manager have to match.

**Comment by mplanchard** [comment]
@exzombie great find, thank you!

Turns out I was on nixpkgs `21.11` and home-manager `21.05`. 

I added `home.enableNixpkgsReleaseCheck = true` to my home-manager config, and then updated my home-manager flakeref to `github:nix-community/home-manager/release-21.11` instead of just `github:nix/community/home-manager/release-21.11`.

This allowed me to specify the `programs.direnv.nix-direnv.enableFlakes` option that @Mic92 mentioned, and once that was enabled, `use flake` in an `.envrc` file worked as expected.
